File: packages/react-native-client/ios/ReactNativeMobileWhepClientViewModule.swift
import ExpoModulesCore
import MobileWhipWhepClient

public class ReactNativeMobileWhepClientViewModule: Module,
    ReconnectionManagerListener
{
    public func onReconnectionStarted() {
        emit(
            event: .reconnectionStatusChanged(
                reconnectionStatus: .reconnectionStarted
            )
        )
    }

    public func onReconnected() {
        emit(event: .reconnectionStatusChanged(reconnectionStatus: .reconnected))
    }

    public func onReconnectionRetriesLimitReached() {
        emit(
            event: .reconnectionStatusChanged(
                reconnectionStatus: .reconnectionRetriesLimitReached
            )
        )
    }

    private func emit(event: WhepEmitableEvent) {
        DispatchQueue.main.async {
            self.sendEvent(event.event.name, event.data)
        }
    }

    struct ConnectOptions: Record {
        @Field
        var serverUrl: String

        @Field
        var authToken: String?
    }

    public func definition() -> ModuleDefinition {
        Name("ReactNativeMobileWhepClientViewModule")

        Events(WhepEmitableEvent.allEvents)

        View(ReactNativeMobileWhepClientView.self) {
            Prop("pipEnabled") {
                (view: ReactNativeMobileWhepClientView, pipEnabled: Bool) in
                view.pipEnabled = pipEnabled
            }
            Prop("autoStartPip") {
                (view: ReactNativeMobileWhepClientView, startAutomatically: Bool) in
                view.autoStartPip = startAutomatically
            }
            Prop("autoStopPip") {
                (view: ReactNativeMobileWhepClientView, stopAutomatically: Bool) in
                view.autoStopPip = stopAutomatically
            }
            Prop("pipSize") { (view: ReactNativeMobileWhepClientView, size: CGSize) in
                view.pipSize = size
            }

            AsyncFunction("createWhepClient") { (
                    view: ReactNativeMobileWhepClientView,
                    configurationOptions: [String: AnyObject]?,
                    preferredVideoCodecs: [String]?,
                    preferredAudioCodecs: [String]?
                ) in
                do {
                    try view.createWhepClient(configurationOptions: configurationOptions, preferredVideoCodecs: preferredVideoCodecs, preferredAudioCodecs: preferredAudioCodecs)
                    view.setReconnectionListener(self)
                    view.setConnectionStateChangeCallback { [weak self] newState in
                        self?.emit(event: .whepPeerConnectionStateChanged(status: newState))
                    }
                } catch {
                    throw Exception(
                        name: "E_WHEP_CLIENT_NOT_CREATED",
                        description:
                            "Creating whep client failed."
                    )
                }
            }

            AsyncFunction("connect") { (view: ReactNativeMobileWhepClientView, connectOptions: ConnectOptions) in

                guard let url = URL(string: connectOptions.serverUrl) else {
                    throw Exception(
                        name: "E_INVALID_SERVER_URL",
                        description: "Invalid server URL. Make sure the address is correct."
                    )
                }

                try await view.connect(serverUrl: url, authToken: connectOptions.authToken)
            }

            AsyncFunction("startPip") { (view: ReactNativeMobileWhepClientView) in
                view.pipController?.startPictureInPicture()
            }

            AsyncFunction("stopPip") { (view: ReactNativeMobileWhepClientView) in
                view.pipController?.stopPictureInPicture()
            }

            AsyncFunction("togglePip") { (view: ReactNativeMobileWhepClientView) in
                view.pipController?.togglePictureInPicture()
            }

            AsyncFunction("disconnect") { (view: ReactNativeMobileWhepClientView) in
                view.disconnect()
            }

            AsyncFunction("pause") { (view: ReactNativeMobileWhepClientView) in
                view.pause()
            }

            AsyncFunction("unpause") { (view: ReactNativeMobileWhepClientView) in
                view.unpause()
            }

            AsyncFunction("getSupportedReceiverVideoCodecsNames") {
                return WhepClient.getSupportedReceiverVideoCodecsNames()
            }

            AsyncFunction("getSupportedReceiverAudioCodecsNames") {
                return WhepClient.getSupportedReceiverAudioCodecsNames()
            }

            AsyncFunction("setPreferredReceiverVideoCodecs") {
                (view: ReactNativeMobileWhepClientView, preferredCodecs: [String]?) in
                view.setPreferredVideoCodecs(
                    preferredCodecs: preferredCodecs
                )
            }

            AsyncFunction("setPreferredReceiverAudioCodecs") {
                (view: ReactNativeMobileWhepClientView, preferredCodecs: [String]?) in
                view.setPreferredAudioCodecs(
                    preferredCodecs: preferredCodecs
                )
            }
        }
    }
}

File: packages/react-native-client/android/src/main/java/com/swmansion/reactnativeclient/ReactNativeMobileWhepClientViewModule.kt
package com.swmansion.reactnativeclient

import android.content.Context
import android.os.Build
import android.util.Log
import android.util.Rational
import com.mobilewhep.client.ClientBaseListener
import com.mobilewhep.client.ClientConnectOptions
import com.mobilewhep.client.ReconnectionManagerListener
import com.mobilewhep.client.WhepClient
import com.mobilewhep.client.WhepConfigurationOptions
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext
import org.webrtc.PeerConnection
import org.webrtc.VideoTrack

class PipSize : Record {
  @Field
  val width: Int = 0

  @Field
  val height: Int = 0
}

class ConnectOptions: Record {
  @Field
  val serverUrl: String = ""

  @Field
  val authToken: String? = null
}

class ReactNativeMobileWhepClientViewModule : Module(), ReconnectionManagerListener {

  interface OnConnectionStateChangeListener {
    suspend fun onConnectionStateChange(newState: PeerConnection.PeerConnectionState)
  }

  fun emit(event: EmitableEvent) {
    sendEvent(event.name, event.data)
  }


  override fun onReconnectionStarted() {
    super.onReconnectionStarted()
    emit(WhepEmitableEvent.reconnectionStatusChanged(ReconnectionStatus.ReconnectionStarted))
  }

  override fun onReconnected() {
    super.onReconnected()
    emit(WhepEmitableEvent.reconnectionStatusChanged(ReconnectionStatus.Reconnected))
  }

  override fun onReconnectionRetriesLimitReached() {
    super.onReconnectionRetriesLimitReached()
    emit(WhepEmitableEvent.reconnectionStatusChanged(ReconnectionStatus.ReconnectionRetriesLimitReached))
  }

  override fun definition() =
    ModuleDefinition {
      Name("ReactNativeMobileWhepClientViewModule")

      Events(WhepEmitableEvent.allEvents)

      View(ReactNativeMobileWhepClientView::class) {
        OnViewDestroys  { view: ReactNativeMobileWhepClientView ->
          runBlocking {
            view.cleanup()
          }
        }

        Prop("pipEnabled") { view: ReactNativeMobileWhepClientView, pipEnabled: Boolean ->
          if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            view.setPictureInPictureEnabled(pipEnabled)
          }
        }

        Prop("autoStartPip") { view: ReactNativeMobileWhepClientView, startAutomatically: Boolean ->
          if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            view.setAutoEnterEnabled(startAutomatically)
          }
        }

        Prop("autoStopPip") { _: ReactNativeMobileWhepClientView, _: Boolean -> }

        Prop("pipSize"){ view: ReactNativeMobileWhepClientView, size: PipSize ->
          if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            view.setAspectRatio(Rational(size.width, size.height))
          }
        }

        AsyncFunction("createWhepClient") { view: ReactNativeMobileWhepClientView, configurationOptions: Map<String, Any>?, preferredVideoCodecs: List<String>?, preferredAudioCodecs: List<String>? ->
          view.createWhepClient(configurationOptions, preferredVideoCodecs, preferredAudioCodecs)
          view.setReconnectionListener(this@ReactNativeMobileWhepClientViewModule)
          view.setConnectionStateChangeListener(object : OnConnectionStateChangeListener {
            override suspend fun onConnectionStateChange(newState: PeerConnection.PeerConnectionState) {
              emit(WhepEmitableEvent.whepPeerConnectionStateChanged(newState))
            }
          })
        }

        AsyncFunction("connect") Coroutine { view: ReactNativeMobileWhepClientView, options: ConnectOptions ->
          view.connect(options)
        }

        AsyncFunction("disconnect") Coroutine { view: ReactNativeMobileWhepClientView ->
          view.disconnect()
        }

        AsyncFunction("pause") { view: ReactNativeMobileWhepClientView ->
          view.pause()
        }

        AsyncFunction("unpause") { view: ReactNativeMobileWhepClientView ->
          view.unpause()
        }

        AsyncFunction("startPip") { view: ReactNativeMobileWhepClientView ->
          if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            view.startPictureInPicture()
          }
        }

        AsyncFunction("stopPip") { _: ReactNativeMobileWhepClientView -> }
        AsyncFunction("togglePip") { _: ReactNativeMobileWhepClientView -> }
      }
    }
}
